### DISCUSSION TITLE
No output with ESP32 S2 Lolin Mini + RPi 5

### DISCUSSION THREAD

**Discussion by justinj57** [awawa-dev/HyperSPI#35] [General]
First off - thanks for all the work you've put into HyperHDR and all the supporting projects.  Fully acknowledge that I'm likely doing something wrong here, but I am pulling my hair out and would be extremely thankful with any help I could get....

Hardware/Software Config
HyperHDR v19
HyperSPI
MS2130 USB Capture Card
3.3V to 5V level shifter
Raspberry Pi 5 (SPI Enabled)
ESP32 S2 Lolin Mini
SK6812 RGBCW
60A 5V Mean Well PSU

I have pins 7, 11, and 12 connected to the corresponding CE0, MOSI, and CLK pins + ground on the Raspberry Pi 5, with ~ 10cm cable.  The ESP32 is powered by USB connection with the RPi5.  I have double and triple checked all connections, made multiple harnesses, etc.  I have also tried multiple ESP32 S2 boards.

I have not been able to get output on GPIO 2.  When I try to run the performance check as outlined in the HyperSPI readme, I get 0 FPS. 

HyperHDR frames: 0 (FPS), receiv.: 0, good: 0, incompl.: 0, mem1: 0, mem2: 0, heap: 193000
RGBW => Gain: 255/255, red: 160, green: 160, blue: 160

 I think this means there's no communicated between the RPi and ESP32, but I am out of ideas on what to check.  Here are the logs from start up to +3 min post start up.  Please let me know if there's other info I could provide that may be helpful in troubleshooting.

Log: https://pastebin.com/Ksx81edA

**Comment by awawa-dev** [comment]
Hi
> I have pins 7, 11, and 12 connected to the corresponding CE0, MOSI, and CLK pins + ground 

Are you sure about it? Because then CE0 and CLK are swapped.
![obraz](https://github.com/awawa-dev/HyperSPI/assets/69086569/de7e3e79-c88e-4094-b3b5-0911251b0250)

Compare theses picture (color of cables are the same).
Closeup of esp32-s2 mini
![https://user-images.githubusercontent.com/69086569/231207350-a670bfea-a96d-4d21-9e8f-f2ca027105da.png](https://user-images.githubusercontent.com/69086569/231207350-a670bfea-a96d-4d21-9e8f-f2ca027105da.png)
Closeup of RPi5 (ignore Pico board)
![288480223-88ccc416-7ca2-43aa-b0da-55ac5993f8f4](https://github.com/awawa-dev/HyperSPI/assets/69086569/f6b5ed8f-0266-4069-a907-00171b40c8ea)

**Comment by justinj57** [comment]
Sorry, I didn't list them in order, but they are connected as such:

ESP32 -> RPi5
7 -> CLK (GPIO 11)
11 ->MOSI (GPIO 10)
12 -> CE0 (GPIO 7)

**Comment by awawa-dev** [comment]
> 12 -> CE0 (GPIO 7)

Check again RPi5 diagram because I think it should be GPIO 8 ...I use only my old photos and the same connector when I need to connect it again so I don't remember exactly.

**Comment by justinj57** [comment]
This is the pinout I was referencing:
<details>
![image](https://github.com/awawa-dev/HyperSPI/assets/70398752/214eaaaa-a36b-4828-ba0a-9cc8820d51fc)
</details>
Should I be using CE1 (GPIO8)?

**Comment by awawa-dev** [comment]
I think this diagram is wrong

**Comment by justinj57** [comment]
Maybe I am looking at it wrong, but the close up pictures appear to be using the same pins that I am using?

**Comment by awawa-dev** [comment]
CE0 is GPIO8 not GPIO7 as the diagram states. For the rest as I mentioned you can compare your connection to mine and also you can post pictures of your setup. Also I dont think default SPI speed is 20000000 but 25000000 in HyperHDR, not sure it is related. But enabling "invert" in the HyperHDR SPI driver properties could be related: disable it.

**Comment by justinj57** [comment]
I messed with the speed and inverting the signal as desperate hopes to get any kind of output.  I can switch them back and provide the logs if that would help?

Here's my current wiring setup:

![20231213_142552](https://github.com/awawa-dev/HyperSPI/assets/70398752/9113a0f2-4f08-4fe4-a0f5-74f9d4bdd2c8)
![20231213_142627](https://github.com/awawa-dev/HyperSPI/assets/70398752/abbf186d-a16a-460e-accc-3e723ddb059e)

**Comment by awawa-dev** [comment]
The connection looks OK. HyperSPI wont work with "invert" option, but non-standard speed should not cause problems for esp32-s2. Try: 25 000 000, 10 000 000 and 1 000 000.

**Comment by awawa-dev** [comment]
BTW, are you using custom firmware with multi-segment enabled? https://github.com/justinj57/HyperSPI/commit/75a00acccc4db104e8c0dc40ec8668616805c189 But in HyperHDR you have an different (and odd) number of LEDs configured, which can lead to unpredictable behavior (worst case scenario: communication works, but esp32-s2 crashes after receiving the first one from the out-of-scenario frame and restarts, therefore the statistics are empty). DSECOND_SEGMENT_START_INDEX is not something that can be set as you like. For testing use the regular released firmware.

**Reply by justinj57** [reply]
Ah - maybe that is it.  I will try the normal firmware here shortly!

**Reply by justinj57** [reply]
I was hopeful, but no change

**Reply by awawa-dev** [reply]
If it doesn't help, diagnosis may be difficult. Try to use a smaller number of LEDs in HyperHDR, e.g. 10, the lowest speed, test without the grabber but with the background effect turned on. If your grabber (ms2130?) doesnt receive video source it produces constant black boards and smoothing may prevent LED rendering if nothing changes (continues output must be enabled). Next, you will need to check the connections in this adapter with a multimeter to see if they are making contact.

**Reply by justinj57** [reply]
Thanks again for your help.

I just noticed in the logs that with a color set, I get this output when switching the LED device on and off:

[PERFORMANCE] [LED0: FPS = 129.11, send = 8266, processed = 10318, dropped = 2051]

But I still get nothing when monitoring the USB serial connection on the RPi with

screen /dev/ttyACM0 115200

Does the log output mean that the SPI connection is active and working?

**Reply by awawa-dev** [reply]
You can test HyperHDR while having `screen` open. If at some point you turn on the LED device in the remote tab, if there is communication, the screen should stop. If you turn off the LED device in the remote, the `screen` should resume displaying messages with updated statistics. I just tested SPI with esp32-s2 and Raspberry Pi 5: it generally works properly, although sometimes there are problems with over 500 LEDs (when 2 frames are sent instead of one, rp2040 also had a problem with this, but I know how to work around it in HyperHDR). It works on all speeds: 1000,000, 10,000,000 and 25,000,000.

**Reply by justinj57** [reply]
Is there a controller you'd recommend over the S2?  I'm willing to try some different hardware at this point

**Reply by awawa-dev** [reply]
Have you tried setting the LED number in HyperHDR to 10 and testing it? It should detect whether there is interference or another problem with SPI communication, the fewer the LEDs, the greater the chance that the frame will be read correctly.

**Reply by justinj57** [reply]
I have, I tried 10 LEDs at a 2000000 rate, still nothing.  Is there any chance it's an issue with the RPi 5?

**Reply by awawa-dev** [reply]
There may be a GPIO fault, as the pins in RPi5 are still not 5V tolerant, although some can survive 5V if special conditions are met. If this is the case, it is difficult to recommend another board for HyperSPI driver because it will not change anything. In the "worst" case, there is still HyperSerialESP32, which in the case of esp32-s2 is not bad because its CDC is very fast and at least you can test your setup. RPi5 also has a second SPI interface available, but I haven't tested it yet

**Reply by awawa-dev** [reply]
To be sure please send me your exported HyperHDR configuration (general tab) that you are using. I will test it on my RPi5. Do you have RPi5 4GB or 8GB and fully updated OS (Bookworm)?

**Reply by awawa-dev** [reply]
Not sure if it can be related https://github.com/raspberrypi/linux/issues/5696 but it should be fixed now

**Reply by justinj57** [reply]
Here's my configuration.  I have the second SPI interface active here, currently trying to see if I can get that to work.  I have a 4GB RPi 5
[HyperHDR-19.0.0.0-Backup-2023-12-14.json](https://github.com/awawa-dev/HyperSPI/files/13675387/HyperHDR-19.0.0.0-Backup-2023-12-14.json)

**Reply by awawa-dev** [reply]
/dev/spidev0.1 is still first interface, only Slave Select is changed to other pin GPIO7 (lest just forget about that wrong diagram posted here earlier). If the first interface is damaged then changing just the slave device wont help almost for sure (unless only GPIO8 is damaged, very unlikely, if it is, it is damaged along with the rest).

**Comment by awawa-dev** [comment]
OK, I restored your configuration.
First of all, I reverted /dev/spidev0.1 to /dev/spidev0.0
In the beginning I had nothing in the statistics like you. Why?
Because there was no video source (I don't have an attached USB grabber).
<img src="https://github.com/awawa-dev/HyperSPI/assets/69086569/5752a1bc-af3a-4780-8a58-0b0516db8967" width="200"/>
I mentioned the background effect earlier. So I activated it:
<img src="https://github.com/awawa-dev/HyperSPI/assets/69086569/914a1d96-06fd-4daa-ba6a-e6eceac06cd3" width="200"/>
Then I had the video source:
<img src="https://github.com/awawa-dev/HyperSPI/assets/69086569/bf72f43a-0eb5-4b4e-a3f2-6fa20d1f79d8" width="200"/>
Also HyperSPI started working from this point. When I stopped the LED device in the Remote HyperHDR tab, the statistics appeared.
![obraz](https://github.com/awawa-dev/HyperSPI/assets/69086569/560390f4-2dac-4f82-b712-7522f37cef61)

**Reply by justinj57** [reply]
Interesting. If you still have this hooked up can you send a picture of the connection? I'm starting to lean towards this being an issue with the RPi

**Comment by awawa-dev** [comment]
> If you still have this hooked up can you send a picture of the connection?

I use the same connection from the beginning
![obraz](https://github.com/awawa-dev/HyperSPI/assets/69086569/66706c4a-853e-4859-88b0-800f28aa9efc)

What OS and kernel version are you using? How did you install it why are you running it on root?

**Reply by justinj57** [reply]
I used the raspberry pi imager running Raspbian 64. I used root as another attempted fix

**Reply by awawa-dev** [reply]
And what is kernel version? Did you always have a video source and read the statistics in `screen` by starting and disabling (at least after few seconds) the LED device in the HyperHDR remote tab?

**Reply by justinj57** [reply]
6.1.0-rpi7-rpi-2712

I was using a solid color (also tried an effect) as the video source. I removed the grabber, as you mentioned without video input it may have been causing issues.  I was turning an color or effect on, enabling the LED device, waiting 15-20 seconds, then disabling.  I never got any performance metrics using this method + screen

**Reply by awawa-dev** [reply]
It looks like you tested correctly. Just wait about 3-5 seconds, but if there is communication, you should notice it almost immediately because the output in the 'screen' console will stop: displaying statistics is stopped if there is any SPI communication received. The color requires "Continue output" because it does not change, the animation effect usually has a fixed output frequency of 25 Hz.

Mine Linux is also 6.1.0-rpi7-rpi-2712.

Before you start checking the next SPI interface, make sure the cables are conducting. Maybe one of them broke in the socket?
Then you can test, for example: /dev/spi1.0 but I haven't checked it yet. In previous RPi 2/3/4 spi0 and spi1 did not have identical capabilities. And the latest RPi5 with a changed GPIO chipset that manages it, was released hastily without completed documentation. You definitely need to activate it somehow in the system first.

**Reply by justinj57** [reply]
I checked 2 sets of cables that I've tried, both are conducting.  Unless I am being completely boneheaded about something, I think this may be a hardware fault.

I went ahead and ordered another RPi to see if that is indeed the issue.  Should be here in a few days, I will replicate the setup again and report back.  Seriously - thanks for the help!

**Reply by awawa-dev** [reply]
1) I found some old program that can test SPI interface.
https://importgeek.wordpress.com/2017/09/11/raspberry-pi-spi-loopback-testing/

you can adjust speed for higher in the code.

2) Also this is my GPIO config (default):
```
pi@rpi5:~/release/build $ pinctrl funcs 0-13
0, ID_SD/GPIO0, SPI0_SIO3, DPI_PCLK, TXD1, SDA0, -, SYS_RIO00, PROC_RIO00, PIO0, SPI2_CE0
1, ID_SC/GPIO1, SPI0_SIO2, DPI_DE, RXD1, SCL0, -, SYS_RIO01, PROC_RIO01, PIO1, SPI2_SIO1
2, PIN3/GPIO2, SPI0_CE3, DPI_VSYNC, CTS1, SDA1, IR_RX0, SYS_RIO02, PROC_RIO02, PIO2, SPI2_SIO0
3, PIN5/GPIO3, SPI0_CE2, DPI_HSYNC, RTS1, SCL1, IR_TX0, SYS_RIO03, PROC_RIO03, PIO3, SPI2_SCLK
4, PIN7/GPIO4, GPCLK0, DPI_D0, TXD2, SDA2, RI0, SYS_RIO04, PROC_RIO04, PIO4, SPI3_CE0
5, PIN29/GPIO5, GPCLK1, DPI_D1, RXD2, SCL2, DTR0, SYS_RIO05, PROC_RIO05, PIO5, SPI3_SIO1
6, PIN31/GPIO6, GPCLK2, DPI_D2, CTS2, SDA3, DCD0, SYS_RIO06, PROC_RIO06, PIO6, SPI3_SIO0
7, PIN26/GPIO7, SPI0_CE1, DPI_D3, RTS2, SCL3, DSR0, SYS_RIO07, PROC_RIO07, PIO7, SPI3_SCLK
8, PIN24/GPIO8, SPI0_CE0, DPI_D4, TXD3, SDA0, -, SYS_RIO08, PROC_RIO08, PIO8, SPI4_CE0
9, PIN21/GPIO9, SPI0_MISO, DPI_D5, RXD3, SCL0, -, SYS_RIO09, PROC_RIO09, PIO9, SPI4_SIO0
10, PIN19/GPIO10, SPI0_MOSI, DPI_D6, CTS3, SDA1, -, SYS_RIO010, PROC_RIO010, PIO10, SPI4_SIO1
11, PIN23/GPIO11, SPI0_SCLK, DPI_D7, RTS3, SCL1, -, SYS_RIO011, PROC_RIO011, PIO11, SPI4_SCLK
12, PIN32/GPIO12, PWM0_CHAN0, DPI_D8, TXD4, SDA2, AAUD_LEFT, SYS_RIO012, PROC_RIO012, PIO12, SPI5_CE0
13, PIN33/GPIO13, PWM0_CHAN1, DPI_D9, RXD4, SCL2, AAUD_RIGHT, SYS_RIO013, PROC_RIO013, PIO13, SPI5_SIO1
```

3) For HyperHDR testing I used this firmware for ESP32-S2 mini (v9 release section):

https://github.com/awawa-dev/HyperSPI/releases/download/v9.0.0.0/hyperspi_esp32_s2_mini_SK6812_RGBW_COLD.bin

**Reply by awawa-dev** [reply]
There is also official SPI loopback tester: https://www.raspberrypi.com/documentation/computers/raspberry-pi.html#troubleshooting-spi-hardware

**Reply by justinj57** [reply]
Well I am completely stumped.  Received the new RPi, used the loopback test and all went well.  I am still getting 0 performance data and no output on pin 2 of the ESP32-S2

**Reply by awawa-dev** [reply]
It's hard for me to tell what could be wrong, did you change the speed in the tester to 25000000? You can also check the eeprom version in RPi: vcgencmd bootloader_version. In the meantime, I tested RPi5 with esp32-s2 using even a 20 cm bundle of ordinary unshielded duponts and it also worked at full speed with less than 500 LEDs. I don't know if it has any significance, for me the ground goes together with the other cables, not separately. When it comes to setup, we use a different fan and USB drive vs SD card, although I don't think it will affect EMI.

**Reply by justinj57** [reply]
Here's the bootloader version:
2023/12/06 18:29:25
version e02d33b3122450accf9dea471a177d3b5623f5ad (release)
timestamp 1701887365
update-time 1703038734
capabilities 0x0000007f

Loopback test worked at 25000000.  

I'm losing my mind - Any chance you'd be willing to test this hardware if I mailed it to you and paid return postage?

**Reply by awawa-dev** [reply]
If you are sure that HyperHDR configuration is correct this time that leaves a cable adapter e.g. crosstalk problem or esp32s2 issue which pinout required soldering. Maybe the time has come to consider flashing HyperSerialESP32 firmware

**Reply by justinj57** [reply]
I wasn't ready to give up yet...

I purchased some new DuPont connectors, since that was the only thing I hadn't replaced yet (made multiple cables, but using the same batch of connectors).  Even though every single cable tested good for continuity, there must have been something screwy/intermittent with the connection to the pins on the Pi or the ESP32, because it now works!

HyperHDR frames: 201 (FPS), receiv.: 201, good: 201, incompl.: 0, mem1: 0, mem2: 0, heap: 192736
RGBW => Gain: 255/255, red: 160, green: 160, blue: 160

Can't tell you how frustrating it is that this was apparently the issue.  Extremely grateful for the help you provided and the work you're doing for the community.

**Reply by awawa-dev** [reply]
I'm glad you didn't give up and managed to run the setup, congratulations! It was an unusual and complicated case. Now if we know that the problem was in the cable connection and at the same time there was continuity, I am almost certain that it was probably caused by unusual ground cabling or crosstalk because of routing of the clock and data signals locked in a separate isolated group without the ground alongside. So far, we could only observe problems when ground was not connected via GPIO: either a complete lack of communication or a very high number of transmission errors. And this despite the fact that the ground of two boards was still connected via USB.

**Comment by justinj57** [comment]
I'll check the eeprom tonight. I did not try other speeds with the loopback but I'll do that as well. Hopefully I can get to the bottom of this
